### PR TITLE
[PF-591] Add remote-safe Harness display-state DTO

### PR DESCRIPTION
**Pull request by mbenhamd**
## Summary
- add a versioned Harness display-state DTO in @mastra/core with JSON-safe encoding for remote session snapshots
- wire server session snapshot responses and schemas to return typed displayState, then regenerate client-js route types
- expose typed RemoteSession.getDisplayState() / HarnessDisplayState in client-js

## Review evidence
- agy review: no blockers; applied toJSON and schema forward-compat recommendations
- local Codex review pass 1: fixed __proto__ preservation in JSON-safe encoding
- local Codex review pass 2: fixed changeset bump, placeholder model tokens, and explicit subagent DTO encoding
- CodeRabbit CLI: applied changeset wording/example findings; did not apply direct server import of the core encoder because check:core-imports fails against the public @mastra/core peer floor for fork-only Harness value imports, so the server keeps a matching local route-layer encoder

## Validation
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/core typecheck
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto build:core
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/core test -- src/harness/v1/display-state.test.ts src/harness/v1/session.display-state.test.ts
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/server check:core-imports
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/server exec vitest run src/server/handlers/harness.test.ts
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/server build:lib
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/client-js test -- src/resources/harness.test.ts
- pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/client-js build:lib
- git diff --check

Linear: PF-591

## ELI5

This PR adds a safe, versioned "snapshot" format for remote Harness sessions so you can take a clean picture of a session's state (what tools are running, queued work, token usage, etc.) and send it over the network without leaking runtime objects or causing serialization issues.

---

## Overview

Adds a versioned Harness display-state DTO (v1) and JSON-safe encoder to capture remote session snapshots, wires that typed snapshot through server responses and client APIs, and provides tests and schemas to validate encoding and shape.

## Key changes

- `@mastra/core`
  - New: packages/core/src/harness/v1/display-state.ts
    - Introduces HarnessDisplayStateSnapshotV1 and related snapshot types (active tool, tool input buffer, active subagent, pending snapshot).
    - Exports HarnessDisplayJsonValue and toHarnessDisplayStateSnapshotV1(state).
    - Encoder rules: Date → ISO, bigint → string, non-finite numbers → null, skip undefined, support toJSON(), protect against cycles (WeakSet), and preserve __proto__ as data (prevent prototype pollution).
  - New tests: packages/core/src/harness/v1/display-state.test.ts — Vitest coverage for encoder behaviors and edge cases.
  - Exports wired: packages/core/src/harness/v1/index.ts re-exports display-state APIs.

- `@mastra/server`
  - New: packages/server/src/server/handlers/harness-display-state.ts
    - Local server-side encoder mirroring core encoding logic (kept local to satisfy `@mastra/core` peer-floor/import constraints).
    - toHarnessDisplayStateSnapshotV1 exported for server use.
  - Updated handlers: packages/server/src/server/handlers/harness.ts
    - Uses server encoder to populate snapshot responses with HarnessDisplayStateSnapshotV1.
    - Replaced import of parseHarnessEventId from core with a local implementation validating harness-v1:<epoch>:<seq>.
  - Schemas: packages/server/src/server/schemas/harness.ts
    - Added harnessDisplayStateSnapshotV1Schema and replaced displayState: z.unknown() with the new typed schema.
  - Tests: packages/server/src/server/handlers/harness.test.ts updated to assert JSON-safe normalization in create/read responses.

- `@mastra/client-js`
  - client-sdks/client-js/src/resources/harness.ts
    - HarnessSessionSnapshot now reintroduces displayState as optional HarnessDisplayStateSnapshotV1.
    - Export alias HarnessDisplayState added.
    - RemoteSession.getDisplayState() now returns HarnessDisplayStateSnapshotV1 | undefined (was unknown).
  - client-sdks/client-js/src/route-types.generated.ts
    - Generated route response types: displayState changed from unknown to the structured union including version, identifiers, lifecycle and nested structured fields (activeTools, toolInputBuffers, activeSubagents, tokenUsage, pending).
    - Nested args/payload/goal types specialized to endpoint-specific auxiliary types.
  - client-sdks/client-js/src/index.ts exports HarnessDisplayState.

- Release & docs
  - .changeset/pf-591-display-state-dto.md: documents minor bumps for core/server/client-js and usage example showing snapshot.version === 1 guard.
  - Commits include refined changeset wording and JSDoc documenting encoder rules.

## Validation & reviews

- Automated/CI validation reported:
  - Typecheck/build for `@mastra/core`; server and client-js builds/tests.
  - Unit tests for core display-state files and server harness handler tests.
  - check:core-imports run and prevented direct server import of core encoder (reason server retains local encoder).
  - git diff --check run.
- Reviews:
  - Agy: no blockers; applied toJSON and schema forward-compat recommendations.
  - Local Codex reviews: fixed __proto__ preservation, changeset/placeholder model token issues, explicit subagent DTO encoding.
  - CodeRabbit CLI: guided changeset wording/examples; flagged core-imports constraint.
- Author followups:
  - Fixed actionable changeset wording and added JSDoc; chose not to manually edit generated route-types file (requested generator-level deduplication in follow-up PR).

## Notes for reviewers / follow-ups

- Server intentionally keeps a local encoder implementation to avoid importing certain Harness runtime types from `@mastra/core` due to peer-floor / fork-only constraints — this is deliberate and validated by check:core-imports.
- Generated route-types file contains duplicated displayState blocks (generated); deduplication should be addressed in the route-types generator in a separate PR if desired.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mbenhamd/mastra/pull/167?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .changeset/pf-591-display-state-dto.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+---
+'@mastra/core': minor
+'@mastra/server': minor
+'@mastra/client-js': minor
+---
+
+Harness v1 remote sessions now let developers reliably inspect remote session state through a versioned `displayState` snapshot (`version: 1`).
+
+The snapshot includes active tools, tool input buffers, active subagents, token usage, pending operations, queue state, and goal metadata without requiring local runtime objects.
+
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply .changeset/pf-591-display-state-dto.md]: Addressed in fca0a57907c by rewriting the PF-591 changeset with action-led wording. Re-ran `git diff --check` and `pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/core test -- src/harness/v1/display-state.test.ts src/harness/v1/session.display-state.test.ts` successfully.

**Comment by mbenhamd** [comment]
dev1/Codex review follow-up after CodeRabbit:

- Fixed the actionable changeset wording comment in `fca0a57907c`.
- Also handled the useful JSDoc nitpick in `92efba078ff` by documenting the JSON-safe display encoder rules.
- I did not hand-edit `client-sdks/client-js/src/route-types.generated.ts` to deduplicate generated displayState blocks. That file is generated from route schemas; manually introducing a shared generic there would drift from the generator output and is out of scope for PF-591. If we want that cleanup, it should be a route-types generator follow-up, not a generated-file patch in this DTO PR.

Validation after the follow-ups:
- `git diff --check`
- `pnpm --dir /tmp/mastra-fork-pf-591-display-state-dto --filter @mastra/core test -- src/harness/v1/display-state.test.ts src/harness/v1/session.display-state.test.ts` — 2 files / 10 tests passed, no type errors.
